test/functional/test-runtime.js

```diff
@@ -307,32 +316,135 @@ describes.fakeWin('runtime', {
         },
       });
       runChunksForTesting(win.document);
-      expect(progress).to.equal('1HIGHA');
+      return promise.then(() => {
+        expect(progress).to.equal('1HIGHA');
+      });
+    });
+  });
 
-      // Runtime mode.
-      win.AMP.push(amp => {
+  it('loads and waits for a single intermediate bundles', () => {
+    // New format: {n:string, f:function(), i: <string|Array<string>}.
+    let progress = '';
+    const queueExtensions = win.AMP;
+    win.AMP.push({
+      n: 'ext1',
+      f: amp => {
         expect(amp).to.equal(win.AMP);
-        progress += '2';
+        progress += 'A';
+      },
+      i: '_base_ext',
+    });
+
+    const promise = adopt(win);
+    const e = Services.extensionsFor(win);
+    const spy = sandbox.spy(e, 'preloadExtension');
+
+    expect(queueExtensions).to.have.length(0);
+    expect(progress).to.equal('');
+    runChunksForTesting(win.document);
+    const script = win.document.querySelector('[data-script=_base_ext]');
+    expect(script).to.be.null;
+    return promise.then(() => {
+      const script = win.document.querySelector('[data-script=_base_ext]');
+      expect(spy).to.have.been.calledWith('_base_ext');
+      expect(script).to.not.be.null;
+      // ext1 should not be executed yet and needs to wait on _base_ext
+      expect(progress).to.equal('');
+      // Manually resolve this extension that doesn't exist in test but which
+      // we tried to load.
+      e.extensions_['_base_ext'].resolve();
+      return e.waitForExtension(win, '_base_ext').then(() => {
+        expect(progress).to.equal('A');
       });
-      win.AMP.push({
-        n: 'ext2',
-        f: amp => {
-          expect(amp).to.equal(win.AMP);
-          progress += 'B';
-        },
+    });
+  });
+
+  it('loads and waits for a multiple intermediate bundles', () => {
+    // New format: {n:string, f:function(), i: <string|Array<string>}.
+    let progress = '';
+    const queueExtensions = win.AMP;
+    win.AMP.push({
+      n: 'ext1',
+      f: amp => {
+        expect(amp).to.equal(win.AMP);
+        progress += 'A';
+      },
+      i: ['_base_ext', '_base_ext2', 'ext0'],
+    });
+
+    win.AMP.push({
+      n: 'ext0',
+      f: amp => {
+        expect(amp).to.equal(win.AMP);
+        progress += 'B';
```

Comment:
So, I think you code actually works, but what you aren't testing yet I think is that dependencies are resolved in order.

I'd expect calls like
```js
win.AMP.push({
     n: '_base_ext',
     f: amp => {
       expect(amp).to.equal(win.AMP);
       progress += 'A';
     },
});
```

Instead of resolving those promises directly.

Reply:
done.